MantidQt/MantidWidgets/src/DataProcessorUI/GenericDataProcessorPresenter.cpp

```diff
@@ -844,108 +841,172 @@ void GenericDataProcessorPresenter::reduceRow(int rowNo) {
+      return;
+    }
+
+    int groupId = m_model->rowCount() - 1;
+    int rowId = m_model->rowCount(m_model->index(groupId, 0));
+
+    // Add a new row to last group
+    insertRow(groupId, rowId);
+  }
   m_tableDirty = true;
 }
 
 /**
-Get the index of the first blank row, or if none exists, returns -1.
+Insert a group after the last selected group
 */
-int GenericDataProcessorPresenter::getBlankRow() {
-  // Go through every column of every row (except for the scale column) and
-  // check if it's blank.
-  // If there's a blank row, return it.
-  const int rowCount = m_model->rowCount();
-  for (int i = 0; i < rowCount; ++i) {
-    bool isBlank = true;
-    for (int j = 0; j < m_columns; ++j) {
-      // Don't bother checking the group column, it'll always have a
-      // value.
-      if (j == m_columns - 2)
-        continue;
-
-      if (!m_model->data(m_model->index(i, j)).toString().isEmpty()) {
-        isBlank = false;
-        break;
-      }
-    }
+void GenericDataProcessorPresenter::appendGroup() {
+  auto selectedGroups = m_view->getSelectedGroups();
 
-    if (isBlank)
-      return i;
+  if (selectedGroups.empty()) {
+    // Append group at the end of the table
+    insertGroup(m_model->rowCount());
+  } else {
+    // Append group after last selected group
+    insertGroup(*(selectedGroups.rbegin()) + 1);
   }
-
-  // There are no blank rows
-  return -1;
+  m_tableDirty = true;
 }
 
 /**
 Delete row(s) from the model
 */
 void GenericDataProcessorPresenter::deleteRow() {
-  std::set<int> rows = m_view->getSelectedRows();
-  for (auto row = rows.rbegin(); row != rows.rend(); ++row)
-    m_model->removeRow(*row);
+  auto selectedRows = m_view->getSelectedRows();
+  for (auto it = selectedRows.rbegin(); it != selectedRows.rend(); ++it) {
+    const int groupId = it->first;
+    auto rows = it->second;
+    for (auto row = rows.rbegin(); row != rows.rend(); ++row) {
+      m_model->removeRow(*row, m_model->index(groupId, 0));
+    }
+  }
+  m_tableDirty = true;
+}
 
+/**
+Delete group(s) from the model
+*/
+void GenericDataProcessorPresenter::deleteGroup() {
+  auto selectedGroups = m_view->getSelectedGroups();
+  for (auto group = selectedGroups.rbegin(); group != selectedGroups.rend();
+       ++group) {
+    m_model->removeRow(*group);
+  }
   m_tableDirty = true;
 }
 
 /**
 Group rows together
 */
 void GenericDataProcessorPresenter::groupRows() {
-  const std::set<int> rows = m_view->getSelectedRows();
-  // Find the first unused group id, ignoring the selected rows
-  const int groupId = getUnusedGroup(rows);
 
-  // Now we just have to set the group id on the selected rows
-  for (auto it = rows.begin(); it != rows.end(); ++it)
-    m_model->setData(m_model->index(*it, m_columns - 2), groupId);
+  // Find if rows belong to the same group
+  // If they do, do nothing
+  // If they don't, remove rows from their groups and add them to a
+  // new group
+
+  auto selectedRows = m_view->getSelectedRows();
```

Comment:
Maybe this could be `const` if the loops below used also `const` iterators / `const` range-based.